src/expr/macro/src/gen.rs

```diff
@@ -328,46 +325,71 @@ impl FunctionAttr {
                 };
             }
         });
-        let let_values = (0..self.args.len()).map(|i| {
-            let v = format_ident!("v{i}");
-            let a = format_ident!("a{i}");
-            quote! { let #v = #a.value_at(row_id); }
-        });
+        let let_values = (0..self.args.len())
+            .map(|i| {
+                let v = format_ident!("v{i}");
+                let a = format_ident!("a{i}");
+                quote! { let #v = unsafe { #a.value_at_unchecked(row_id) }; }
+            })
+            .collect_vec();
         let let_state = match &self.state {
-            Some(_) => quote! { self.state.take() },
-            None => quote! { self.state.as_ref().map(|x| x.as_scalar_ref()) },
+            Some(s) if s == "ref" => quote! { self.state.as_ref().map(|x| x.as_scalar_ref()) },
+            _ => quote! { self.state.take() },
         };
         let assign_state = match &self.state {
-            Some(_) => quote! { state },
-            None => quote! { state.map(|x| x.to_owned_scalar()) },
+            Some(s) if s == "ref" => quote! { state.map(|x| x.to_owned_scalar()) },
+            _ => quote! { state },
         };
         let init_state = match &self.init_state {
-            Some(s) => s.parse().unwrap(),
+            Some(s) => format!("Some({s})").parse().unwrap(),
             _ => quote! { None },
         };
         let fn_name = format_ident!("{}", self.user_fn.name);
-        let mut next_state = quote! { #fn_name(state, #args) };
+        let args = (0..self.args.len()).map(|i| format_ident!("v{i}"));
+        let args = quote! { #(#args,)* };
+        let retract = match self.user_fn.retract {
+            true => quote! { matches!(input.ops()[row_id], Op::Delete | Op::UpdateDelete) },
```

Comment:
What about using `get_unchecked` here?

Reply:
This does help. 👍

  | ops[idx] | ops.get_unchecked | Cmp
-- | -- | -- | --
count_agg_vis_rate_0_05 | 152 | 146 | -4%
count_agg_vis_rate_0_25 | 462 | 443 | -4%
count_agg_vis_rate_0_5 | 905 | 858 | -5%
count_agg_vis_rate_0_75 | 1,279 | 1,187 | -7%
count_agg_without_vis | 949 | 950 | 0%
max_agg_vis_rate_0_05 | 162 | 156 | -4%
max_agg_vis_rate_0_25 | 510 | 490 | -4%
max_agg_vis_rate_0_5 | 1,001 | 926 | -7%
max_agg_vis_rate_0_75 | 1,425 | 1,306 | -8%
max_agg_without_vis | 1,080 | 1,080 | 0%
min_agg_vis_rate_0_05 | 162 | 156 | -4%
min_agg_vis_rate_0_25 | 526 | 471 | -10%
min_agg_vis_rate_0_5 | 1,017 | 926 | -9%
min_agg_vis_rate_0_75 | 1,419 | 1,305 | -8%
min_agg_without_vis | 1,106 | 1,081 | -2%
sum_agg_vis_rate_0_05 | 253 | 244 | -4%
sum_agg_vis_rate_0_25 | 929 | 858 | -8%
sum_agg_vis_rate_0_5 | 1,801 | 1,671 | -7%
sum_agg_vis_rate_0_75 | 2,886 | 2,598 | -10%
sum_agg_without_vis | 3,093 | 2,559 | -17%

Reply:
Cool!